jetty-server/src/main/java/org/eclipse/jetty/server/ContentProcessor.java

```diff
@@ -0,0 +1,205 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.Objects;
+
+import org.eclipse.jetty.util.thread.AutoLock;
+
+public abstract class ContentProcessor extends Content.Processor
+{
+    private final AutoLock _lock = new AutoLock();
+    private boolean _reading;
+    private boolean _iterating;
+    private boolean _available;
+    private Content _output;
+    private Runnable _onContentAvailable;
+
+    public ContentProcessor(Content.Provider provider)
+    {
+        super(provider);
+    }
+
+    @Override
+    public Content readContent()
+    {
+        boolean available;
+        try (AutoLock ignored = _lock.lock())
+        {
+            if (_reading)
+                return null;
+
+            if (_output != null)
+            {
+                Content output = _output;
+                _output = null;
+                return output;
+            }
+
+            available = _available;
+            _available = false;
+            _reading = true;
+        }
+
+        try
+        {
+            while (true)
+            {
+                Content output = available ? null : produce(null);
```

Comment:
`produce(null)` should be avoided (also, not sure the logic above is correct for the first pass -- available seems to either be `false` or skip `produce()`?).

If upstream always has content, `produce(null)` should never happen.
I would prefer the contract of `produce()` to be fully consuming the input content, and return the first transformed.
Method `readContent()` will take care of returning transformed content if present; if not present it should demand for it, similar to HTTP/1 `fillAndParse()` -- we only return `null` from `readContent()` if we tried and we are truly stalled.

Reply:
> If upstream always has content, `produce(null)` should never happen.

That was the point of the discussion I tried to have in the issue.   It may be that upstream has content, but the previous content is still able to produce more content (eg being expanded), so we need to call the produce again when there is more demand to see if it can produce more.   There are designs that don't need to pass null, but they have other issues (like letting the produce go beyond demand).

> I would prefer the contract of produce() to be fully consuming the input content,
Because we are using `Content` with `release()` rather than just `ByteBuffers`, then we cannot use consumption as a signal for anything.   A Processor is free to hold onto a content over multiple cycles and release it when it wants.  It is also free to pass it onto the application, which may then hold onto it for a while until it calls release.  The whole point of `release()` is to avoid copying (ie forcing consumption before it is necessary).       If the ultimate target of this chain is a servlet application reading 1 byte at a time, we don't want to force anybody to fully consume content as that will imply a needless copy and buffering.

As far as I can tell, the state of consumption of a content can have no influence on how we process it.   Even a consumed content might be retained so that the processor or application can pick bytes out of it later.

> Method readContent() will take care of returning transformed content if present;

If we go with that approach, then it may be that a processor fully consumes some content that arrives.  In that case, we will have called the consumers/applications onContentAvailable method and it will do a `readContent` that returns null.     That may be OK, but we'd have to rename the concept from `onContentAvailable` to 'tryAgainNowButItMayStillBeNull` (or something better to that intent).       I think the impl I have here shows that it is at least possible to "filter" content availability and to only call the consumer/application once there really is content available.  Cost is we have to "buffer" one content between when notification and a subsequent call to `readContent()`.